File: flink-app/src/main/java/com/example/ctr/application/CtrJobService.java
package com.example.ctr.application;

import com.example.ctr.infrastructure.flink.CtrJobPipelineBuilder;
import com.example.ctr.infrastructure.flink.FlinkEnvironmentFactory;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

/**
 * CTR Flink 잡을 시작하는 서비스.
 *
 * 책임:
 * 1️⃣ {@link FlinkEnvironmentFactory} 를 통해 설정된
 * {@link StreamExecutionEnvironment} 획득
 * 2️⃣ {@link CtrJobPipelineBuilder} 로 데이터 파이프라인 구성
 * 3️⃣ 잡 실행 및 그레이스풀 셧다운 처리
 */
@Slf4j
@RequiredArgsConstructor
public class CtrJobService {

    private final FlinkEnvironmentFactory envFactory;
    private final CtrJobPipelineBuilder pipelineBuilder;
    private volatile StreamExecutionEnvironment env;

    public void execute() {
        try {
            log.info("Starting CTR Calculator Flink Job (Exactly‑once)...");

            // 실행 환경 생성 (체크포인트/재시작 전략 설정 등)
            env = envFactory.create();

            // 파이프라인 구성 (소스, 집계, 싱크)
            pipelineBuilder.build(env);

            // 그레이스풀 셧다운을 위한 훅 등록
            registerShutdownHook();

            log.info("Executing Flink job...");
            env.execute("CTR Calculator Job (Exactly‑once)");
            log.info("Flink job completed successfully");
        } catch (Exception e) {
            log.error("Failed to execute Flink job", e);
            throw new RuntimeException("Flink job execution failed", e);
        }
    }

    private void registerShutdownHook() {
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            log.info("Shutdown hook triggered. Gracefully stopping Flink job...");
            try {
                if (env != null) {
                    // JVM 종료 시 Flink가 그레이스풀 셧다운을 처리
                    log.info("Flink job shutdown initiated");
                }
            } catch (Exception e) {
                log.error("Error during graceful shutdown", e);
            }
        }, "flink-shutdown-hook"));
    }

}

File: flink-app/src/main/kotlin/com/example/ctr/application/CtrJobService.kt
package com.example.ctr.application

import com.example.ctr.infrastructure.flink.CtrJobPipelineBuilder
import com.example.ctr.infrastructure.flink.FlinkEnvironmentFactory
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment
import org.slf4j.LoggerFactory

class CtrJobService(
    private val envFactory: FlinkEnvironmentFactory,
    private val pipelineBuilder: CtrJobPipelineBuilder
) {

    private val log = LoggerFactory.getLogger(CtrJobService::class.java)
    @Volatile
    private var env: StreamExecutionEnvironment? = null

    fun execute() {
        try {
            log.info("Starting CTR Calculator Flink Job (Exactly-once)...")

            val executionEnv = envFactory.create()
            env = executionEnv
            pipelineBuilder.build(executionEnv)
            registerShutdownHook()

            log.info("Executing Flink job...")
            executionEnv.execute("CTR Calculator Job (Exactly-once)")
            log.info("Flink job completed successfully")
        } catch (e: Exception) {
            log.error("Failed to execute Flink job", e)
            throw RuntimeException("Flink job execution failed", e)
        }
    }

    private fun registerShutdownHook() {
        Runtime.getRuntime().addShutdownHook(Thread({
            log.info("Shutdown hook triggered. Gracefully stopping Flink job...")
            try {
                env?.let {
                    log.info("Flink job shutdown initiated")
                }
            } catch (e: Exception) {
                log.error("Error during graceful shutdown", e)
            }
        }, "flink-shutdown-hook"))
    }
}
